sqlrunner/main.py

```diff
@@ -94,12 +97,23 @@ def run_sql(*, dsn, sql_query, include_statistics=False):
     conn.close()
 
 
-def write_results(results, f_path):
-    # job-runner expects the output CSV file to exist. If it doesn't, then the SQL
-    # Runner action will fail. A user won't know whether their query returns any
-    # results, so to avoid the SQL Runner action failing, we write an empty output CSV
-    # file.
-    utils.touch(f_path)
+def write_results(results, destination):
+    dest_is_path = isinstance(destination, pathlib.Path)
+
+    if dest_is_path:
+        # If writing to file, job-runner expects the output CSV file to exist.
+        # If it doesn't, then the SQL Runner action will fail.
+        # A user won't know whether their query returns any results,
+        # so to avoid the SQL Runner action failing,
+        # we write an empty output CSV file.
+        utils.touch(destination)
+
+        if destination.suffixes == [".csv", ".gz"]:
+            destination = gzip.open(
+                destination, "wt", newline="", encoding="utf-8", compresslevel=6
+            )
+        else:
+            destination = destination.open(mode="w", newline="", encoding="utf-8")
```

Comment:
If you wanted to simplify the signature of `write_results`, then you could move this into a function that returns a file-like object and call that function in `main`.

Reply:
Yes, it could be but there's a bit more complexity around closing the file-like object and not closing stdout

Reply:
I was wondering whether we could tidy away this complexity with a custom context manager. Not essential, though.

Reply:
In my now-deleted first draft I decorated `sys.stdout` such that it had a no-op `close()` I'm not sure either is worth the effort. I guess with a custom context manager you could do the file initialisation/opening logic as well